src/main/java/org/radarcns/management/service/MetaTokenService.java

```diff
@@ -0,0 +1,135 @@
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Created by nivethika.
+ *
+ * <p>Service to delegate MetaToken handling.</p>
+ *
+ */
+@Service
+@Transactional
+public class MetaTokenService {
+
+    private final Logger log = LoggerFactory.getLogger(MetaTokenService.class);
+
+    @Autowired
+    private MetaTokenRepository metaTokenRepository;
+
+    @Autowired
+    private ManagementPortalProperties managementPortalProperties;
+
+    /**
+     * Save a metaToken.
+     *
+     * @param metaToken the entity to save
+     * @return the persisted entity
+     */
+    public MetaToken save(MetaToken metaToken) {
+        log.debug("Request to save MetaToken : {}", metaToken);
+        return metaTokenRepository.save(metaToken);
+    }
+
+    /**
+     * Get one project by id.
+     *
+     * @param tokenName the id of the entity
+     * @return the entity
+     * @throws CustomNotFoundException if there is no project with the given id
+     */
+    public TokenDTO fetchToken(String tokenName) throws
+            MalformedURLException, CustomNotFoundException {
+        log.debug("Request to get Token : {}", tokenName);
+        MetaToken fetchedToken = getToken(tokenName);
+
+        // process the response if the token is not fetched or not expired
+        if (!fetchedToken.isFetched() && Instant.now().isBefore(fetchedToken.getExpiryDate())) {
+            // create response
+            TokenDTO result = new TokenDTO(fetchedToken.getToken(),
+                    new URL(managementPortalProperties.getCommon().getBaseUrl()));
+            // change fetched status to true.
+            fetchedToken.fetched(true);
+            save(fetchedToken);
+            return result;
+        } else {
+            throw new CustomParameterizedException("invalidRequest", "token is already fetched "
+                + "or has expired. Invalid request");
+        }
+
+
+    }
+
+    /**
+     * Gets a token from databased using the tokenName.
+     *
+     * @param tokenName tokenName.
+     * @throws CustomNotFoundException if the token not found.
+     * @return fetched token as {@link MetaToken}.
+     */
+    @Transactional(readOnly = true)
+    public MetaToken getToken(String tokenName) throws CustomNotFoundException {
+        Optional<MetaToken> fetchedToken = metaTokenRepository.findOneByTokenName(tokenName);
+
+        if (fetchedToken.isPresent()) {
+            return  fetchedToken.get();
+        } else {
+            throw  new CustomNotFoundException(
+                ErrorConstants.ERR_TOKEN_NOT_FOUND,
+                Collections.singletonMap("tokenName", tokenName));
+        }
+    }
+
+    /**
+     * Builds a unique meta-token instance, by checking for token-name collision.
+     * @return an unique token
+     */
+    public MetaToken buildUniqueToken() {
```

Comment:
Because the search for a unique token is not atomic with saving a token, this does not guarantee a unique token. Flow: process 1 creates token, process 2 creates identical token, both are stored one after the other. The second `save` call should give a conflict from the DB, because there is a unique constraint. Ideally, buildUniqueToken would also store the token and solve the conflict by catching an exception and then choosing a new token name, instead of searching. Or catch it one layer up, in the `/oauth-clients/pair` request handling. This is also faster than searching the repository, because it is an exceptional situation. So in most situations, this check is going to return false, but it still costs some overhead.

Reply:
Ok. Would creating identical token includes the `id` of the token or should I skip it so it focuses on the `tokenName` unique constraint?

Reply:
The ID would be autogenerated, right? So it will always be unique. But the tokenName is generated by this code, so we need to resolve conflicts there as well.

Reply:
Yes. ID is auto-generated.